### PR TITLE
Eliminated JVM crashing race on transport stop

### LINKED ISSUE #21336: Close the media driver only once
As doing it twice crashes the JVM.

### PR DESCRIPTION
**Pull request by johanandren**
Fixes #21336

### CODE REVIEW

**Comment by patriknw** [comment]
LGTM, nice detective work

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 1]
```diff
@@ -6,10 +6,9 @@ package akka.remote.artery
 import java.io.File
 import java.net.InetSocketAddress
 import java.nio.channels.{ DatagramChannel, FileChannel }
-
 import java.util.concurrent.CopyOnWriteArrayList
 import java.util.concurrent.TimeUnit
-import java.util.concurrent.atomic.AtomicLong
+import java.util.concurrent.atomic.{ AtomicBoolean, AtomicLong, AtomicReference }
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: you don't use the Boolean anymore?

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 2]
```diff
@@ -438,19 +436,22 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
       log.debug("Started embedded media driver in directory [{}]", driver.aeronDirectoryName)
       topLevelFREvents.loFreq(Transport_MediaDriverStarted, driver.aeronDirectoryName().getBytes("US-ASCII"))
       Runtime.getRuntime.addShutdownHook(stopMediaDriverShutdownHook)
-      mediaDriver = Some(driver)
+      mediaDriver.set(Some(driver))
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: make it a CAS and fail if it was set before. Unlikely but better fail than having two drivers concurrently or something even more weirder.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala [hunk 3]
```diff
@@ -438,19 +436,22 @@ private[remote] class ArteryTransport(_system: ExtendedActorSystem, _provider: R
...
     }
   }
 
-  private def aeronDir: String = mediaDriver match {
+  private def aeronDir: String = mediaDriver.get match {
```
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: does it make sense to cache this? Or this is used rarely?

**Comment by drewhk** [comment]
LGTM

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 2 above)
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: You only get here through `ArteryTransport.start` and I'm pretty sure there will all kinds of catastrophy that already happened if you do that more than once.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 3 above)
Comment (johanandren) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: Two places during transport startup, so probably not worth caching.

File: akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala (hunk 2 above)
Comment (drewhk) [inline akka-remote/src/main/scala/akka/remote/artery/ArteryTransport.scala]: It does not hurt to cry out an explicit error message though?

**Comment by ktoso** [comment]
LGTM, curious that such a thing caused crashing.

**Pull request merged**